### PR TITLE
Rdm 4109 BE changes to allow: Management web consuming the new payment history base type

### PR DESCRIPTION
**Pull request by andrewfolga**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-4109

### Change description ###
BE changes to allow: Management web consuming the new payment history base type

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java [hunk 1]
```diff
@@ -85,4 +102,18 @@ CaseTabCollection getCaseTabCollection(String caseTypeId) {
             .collect(Collectors.toList());
     }
 
+    void hydrateHistoryField(CaseDetails caseDetails, CaseType caseType, List<CaseViewEvent> events) {
+        final List<CaseField> caseFieldDefinitions = caseType.getCaseFields();
+        final JsonNode data = MAPPER.convertValue(caseDetails.getData(), JsonNode.class);
+        if (data != null) {
```
Comment (rafalkalita) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java]: What is the use of data variable? It is not clear.

Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java]: This is to build a value for a field of type CaseHistoryViewer. This is making the case history tab a flexible control.

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseTabCollection.java [hunk 2]
```diff
@@ -36,4 +36,16 @@ public void setChannels(List<String> channels) {
...
+
+    public boolean hasTabFieldType(String tabFieldType) {
+        return this.tabs.stream()
+                 .flatMap(tab -> tab.getTabFields().stream()
+                                   .filter(field -> field
+                                       .getCaseField()
+                                       .getFieldType()
+                                       .getType()
+                                       .equals(tabFieldType)))
+                 .findAny()
```
Comment (gbenadikar) [inline src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseTabCollection.java]: anyMatch() would have been more readable

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java [hunk 3]
```diff
@@ -1,23 +1,36 @@
...
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static uk.gov.hmcts.ccd.domain.model.definition.FieldType.CASE_HISTORY_VIEWER;
+
 public abstract class AbstractDefaultGetCaseViewOperation {
 
     public static final String QUALIFIER = "default";
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
Comment (gbenadikar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java]: Should avoid creating instances of objectMapper throughout the applicable. There is a ObjectMapperService that should be reused or any feature not present there should be added to that service.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewOperationTest.java [hunk 4]
```diff
@@ -1,36 +1,65 @@
...
+import uk.gov.hmcts.ccd.domain.model.definition.FieldType;
+import uk.gov.hmcts.ccd.domain.model.definition.Jurisdiction;
 import uk.gov.hmcts.ccd.domain.model.std.AuditEvent;
 import uk.gov.hmcts.ccd.domain.service.common.CaseTypeService;
 import uk.gov.hmcts.ccd.domain.service.common.EventTriggerService;
 import uk.gov.hmcts.ccd.domain.service.common.UIDService;
 import uk.gov.hmcts.ccd.domain.service.getcase.GetCaseOperation;
 import uk.gov.hmcts.ccd.domain.service.getevents.GetEventsOperation;
 
+import java.time.LocalDateTime;
```
Comment (gbenadikar) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewOperationTest.java]: I think the java imports are at the beginning, followed by static imports and then other classes. Checkstyle may complain.

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java [hunk 5]
```diff
@@ -1188,11 +1193,63 @@ public CaseViewEvent build() {
         }
     }
 
+    public static class CaseTypeTabFieldBuilder {
+        private final CaseTypeTabField caseTypeTabField;
+
+        private CaseTypeTabFieldBuilder()
+        {
```
Comment (gbenadikar) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: checkstyle error on open brace

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java [hunk 6]
```diff
@@ -1188,11 +1193,63 @@ public CaseViewEvent build() {
...
+        }
+
+    }
+
+    public static class CaseTypeTabBuilder {
+        private final CaseTypeTab caseTypeTab;
+        private final List<CaseTypeTabField> caseTypeTabFields;
+
+        private CaseTypeTabBuilder()
+        {
```
Comment (gbenadikar) [inline src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: checkstyle

File: src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseTabCollection.java (hunk 2 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/model/definition/CaseTabCollection.java]: Done

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java (hunk 3 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java]: I agree to that but we have other places that create their own mapper too. 🤔

File: src/test/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewOperationTest.java (hunk 4 above)
Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/aggregated/DefaultGetCaseViewOperationTest.java]: Done

File: src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java (hunk 3 above)
Comment (andrewfolga) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/aggregated/AbstractDefaultGetCaseViewOperation.java]: Done

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java (hunk 5 above)
Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: Done

File: src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java (hunk 6 above)
Comment (andrewfolga) [reply src/test/java/uk/gov/hmcts/ccd/domain/service/common/TestBuildersUtil.java]: Done

**Pull request merged**